src/bluesearch/entrypoint/database/convert_pdf.py

```diff
@@ -124,38 +145,124 @@ def run(
     int
         The exit code of the command
     """
-    # Check if the input file exists
-    if not input_pdf_path.exists():
-        logger.error(
-            f"The input file {str(input_pdf_path)!r} does not exist.",
-        )
+    # Check the GROBID server
+    if not grobid_is_alive(grobid_host, grobid_port):
+        logger.error("The GROBID server is not alive")
         return 1
 
-    # Check if the output file already exists
-    if output_xml_path.exists() and not force:
-        logger.error(
-            f"The output file {str(output_xml_path)!r} already exists. "
-            "Either delete it or use the --force option to overwrite it.",
-        )
+    # Check if the input file exists
+    if not input_path.exists():
+        logger.error(f"The input path {str(input_path)!r} does not exist")
         return 1
 
-    # Read the PDF file
-    logger.info("Reading the PDF file")
-    with input_pdf_path.open("rb") as fh_pdf:
-        pdf_content = fh_pdf.read()
+    # Collect input paths
+    input_paths: Iterable[pathlib.Path]
+    if input_path.is_file():
+        input_paths = [input_path]
+    else:
+        input_paths = input_path.rglob("*.pdf")
 
-    # Convert the PDF to XML
-    logger.info("Converting PDF to XML")
-    from bluesearch.database.pdf import grobid_pdf_to_tei_xml
+    failed_paths = []
+    path_map = _prepare_output_paths(input_paths, output_dir, force)
 
-    xml_content = grobid_pdf_to_tei_xml(pdf_content, grobid_host, grobid_port)
+    if len(path_map) == 0:
+        logger.warning("No files to process, stopping")
+        return 0
 
-    # Write the XML file
-    logger.info("Writing the XML file to disk")
-    with output_xml_path.open("w") as fh_xml:
-        n_bytes = fh_xml.write(xml_content)
-    logger.info("Wrote %d bytes to %s", n_bytes, output_xml_path.resolve().as_uri())
+    if output_dir is not None:
+        output_dir.mkdir(exist_ok=True)
+
+    # Convert
+    def do_work(pdf_path):
+        xml_path = path_map[pdf_path]
```

Comment:
That might not be always thread-safe ([doc](https://docs.python.org/3/glossary.html#term-global-interpreter-lock)).

Reply:
Yes, we did think about the thread-safety when writing this.

The article you linked is just the definition of the GIL, but I'm not seeing anything about thread-safety of lists/dicts there, or did I oversee something? Wouldn't the existence of GIL actually not make the opposite more likely?

Anyway, when we tried to research the same question we came across this [SO thread](https://stackoverflow.com/questions/6319207/are-lists-thread-safe). There, people seem to imply that the two operations we do (reading a value from a dict and appending a value to a list) are thread-safe. What do you think?

Reply:
Both the documentation I have linked and the SO thread say that the thread-safety is only true on CPython. So, if someone use another implementation, for performances for example, this code will no more be thread-safe.

Reply:
I pushed a couple of new commits so that there should no longer be any thread safety concern.
Does the code now look fine after 5725821?